File: devide and conquer/MergeSort.c
//merge sort algorithm
//this is a sorting algorithm

#include <stdio.h>
#define n 7

void merge(int a[],int beg,int mid,int end){//this is for sorting and merging back the array
    int i,j,k;
    int n1= mid-beg+1;
    int n2 = end-mid;

    int LeftArray[n1],RightArray[n2];

    for(int i=0;i<n1;i++){
        LeftArray[i] = a[beg+i];
    }
    for(int j=0;j<n2;j++){
        RightArray[j]=a[mid+1+j];
    }

    i=0;
    j=0;
    k=beg;

    while(i<n1&&j<n2){
        if(LeftArray[i]<=RightArray[j]){
            a[k]=LeftArray[i];
            i++;
        }
        else{
            a[k]=RightArray[j];
            j++;
        }
        k++;
    }
    while(i<n1){
        a[k]=LeftArray[i];
        i++;
        k++;
    }
    while(j<n2){
        a[k]=RightArray[j];
        j++;
        k++;
    }
}
void mergeSort(int a[],int beg,int end){//this is for devide and conquer
    if(beg<end){
        int mid = (beg+end)/2;
        mergeSort(a,beg,mid);
        mergeSort(a,mid+1,end);
        merge(a,beg,mid,end);
    }
}
void printArr(int a[]){
    for(int i=0;i<n;i++){
        printf("%d\t",a[i]);
    }
    printf("\n");
}
int main(){
    int array[n];
    printf("insert element in the array \n");
    for(int i=0;i<n;i++){
        scanf("%d",&array[i]);
    }
    printf("array before sorting \n");
    printArr(array);
    mergeSort(array,0,n-1);
    printf("array elements after sorting \n");
    printArr(array);
}

File: devide and conquer/MergeSort.java
//this is a merge sort program
//this uses devide and conquer technique
/**
    merge sort has the time complexity of :- o(nlogn) in worst case
 */

public class MergeSort{
    public static void Merge(int[] array,int low,int mid,int high){
        int i = low;
        int j = mid+1;
        int k = low;
        int[] temp = new int[array.length];
        while(i<=mid&&j<=high){
            if(array[i]<array[j]){
               temp[k] = array[i];
               i++; 
            }
            else{
                temp[k] = array[j];
                j++;
            }
            k++;
        }
        if(i>mid){
            while(j<=high){
                temp[k]=array[j];
                j++;
                k++;
            }
        }
        else{
            while(i<=mid){
                temp[k]=array[i];
                i++;
                k++;
            }
        }
        for(k=low;k<=high;k++){
            array[k] = temp[k];
        }
    }
    public static void MergeSort(int[] array,int low,int high){
        if(low<high){
            int mid = (low+high)/2;
            MergeSort(array,low,mid);
            MergeSort(array,mid+1,high);
            Merge(array,low,mid,high);
        }
    }
    public static void main(String[] args){
        int[] array = {10,2,15,1,25,40};
        MergeSort(array,0,array.length-1);
    }
}